Adapt user/integration guide according to new flow with adyen-web-components (#329)
commercetools/commercetools-adyen-integration: extension/docs/WebComponentsIntegrationGuide.md

<!-- START doctoc generated TOC please keep comment here to allow auto update -->

> <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/04edc7df1187eba2bc4144eb4bd3d7f96ac79e11

  - [Step 2: Get available payment methods](#step-2-get-available-payment-methods)
  - [Step 3: Add Components to your payments form](#step-3-add-components-to-your-payments-form)
  - [Step 4: Make a payment](#step-4-make-a-payment)
      - [Klarna payment](#klarna-payment)
  - [Step 5: Submit additional payment details](#step-5-submit-additional-payment-details)
  - [Step 6: Capture payment (required for Klarna)](#step-6-capture-payment-required-for-klarna)
  - [Error handling](#error-handling)
    - [Extension module errors](#extension-module-errors)
    - [Adyen payment refusals](#adyen-payment-refusals)
    - [Shopper successfully paid but `redirectUrl` was not reached](#shopper-successfully-paid-but-redirecturl-was-not-reached)
    - [Shopper tries to pay a different amount than the actual order amount](#shopper-tries-to-pay-a-different-amount-than-the-actual-order-amount)
  - [Test and go live](#test-and-go-live)
- [Manual Capture](#manual-capture)
- [Cancel or refund](#cancel-or-refund)
- [Bad Practices](#bad-practices)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Web Components integration guide

> In this integration process, there are different parties working with the extension module. Those: 

Author: I am not sure about that, does stakeholder correct term in here for a server?

> In this integration process, there are different parties working with the extension module. Those: 

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/8de846c99224cd4de134ec08cbc305839dbcf474

- **Shopper** - a person that's using the shop.
- **Browser** - frontend part of the checkout UI (webshop). 
- **Merchant server** - backend part of the checkout.
- **Extension module** - extension module is handling checkout steps integrating with Adyen and CTP payment object API calls with the [API Extensions](https://docs.commercetools.com/http-api-projects-api-extensions).
- **Notification module** - [notification module](./../../notification/README.md) is receiving notifications from Adyen, processing and storing them on a commercetools payment object.

The following diagram shows the whole checkout integration flow using with [Adyen Web Components](https://docs.adyen.com/checkout/components-web).

> ![Flow](https://user-images.githubusercontent.com/3469524/85017686-3317bf00-b16c-11ea-8840-f34b97ac3dcb.jpeg)

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/4d33229bff06dd92bdaec5ab7043f590e0965bf1
https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/26ac50aacc5391721d4d756ce173956af15b4b98
https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/4e415a0fcf8cf7b9dd6bcf6461947dcca540110d

## Before you begin

> If you haven't done so already, follow the official Adyen [get started guide](https://docs.adyen.com/checkout/get-started) to set up your test account, get your API key.

Author: 48f877510451939fc5beb6fb3ae260cb62f0a25c

In order to make the extension module working, follow our [deployment guide](./DeploymentGuide.md).

> ## Step 1: CTP checkout validations

Author: There are 6 steps, steps can not be omitted.
https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/33a36ecb3cd4609312be91227890763844f8a448

> In the merchant server, ensure the steps below are done:

Author: It's described above as `Merchant server - backend part of the checkout.`

> In the merchant server, ensure the steps below are done:

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/44c643e7d0c40dd563b8459f33743b1514c4c72e

If shopper decides to change cart's amount within the second tab and finalize payment within the first tab, then according to payment amount validation an error
will be shown and order creation declined.

## Test and go live
After doing the all steps above and testing the payment methods with your test accounts, then when you are ready to go live, you need to do the following steps described as [testing the integration](https://docs.adyen.com/checkout/components-web/#testing-your-integration).

Additionally, follow the official Adyen [integration checklist](https://docs.adyen.com/development-resources/integration-checklist) to ensure you have a complete implementation into Adyen.

# Manual Capture
By default, payments are captured immediately after authorisation. For payment methods that support separate authorisation and capture, you also have the option to capture the payment later, for example only after the goods have been shipped. This allows you to cancel the payment in case of any issues with the shipment. 

If you need to explicitly request a capture for each payment please follow our [manual capture documentation](./ManualCapture.md).

# Cancel or refund

> If you want to return the funds to your shopper, but are not certain whether the payment has been captured or not, use the Cancel or Refund functionality. 

Reviewer: what are the steps to follow if the payment has been captured for sure?

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/502d2fac7105d8bce95ef463227c31719aecb875

  - [Test and go live](#test-and-go-live)
- [Manual Capture](#manual-capture)
- [Cancel or refund](#cancel-or-refund)
- [Bad Practices](#bad-practices)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Web Components integration guide

Terms used in this guide:

- **Shopper** - a person that's using the shop.
- **Browser** - frontend part of the checkout UI (web shop). 
- **Merchant server** - backend part of the checkout.
- **Extension module** - extension module is handling checkout steps integrating with Adyen and commercetools payment object HTTP API calls with the [commercetools HTTP API Extensions](https://docs.commercetools.com/http-api-projects-api-extensions).
- **Notification module** - [notification module](./../../notification/README.md) is receiving notifications from Adyen, processing and storing them on a commercetools payment object.

The following diagram shows the whole checkout integration flow using with [Adyen Web Components](https://docs.adyen.com/checkout/components-web).

![Flow](https://user-images.githubusercontent.com/3469524/86238652-2c427080-bb9e-11ea-83cf-246bbc3a23f3.png)

## How it works

> On this guide we describe both server-side and client-side in the checkout integration **steps**:

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/c5e0ddba92bdef3c4376c6d3198d5dbacdb322c3

- [Step 1](#step-1-commercetools-checkout-validations) : Do the required checkout validations in your merchant server.
- [Step 2](#step-2-get-available-payment-methods): Set `getPaymentMethodsRequest` custom field to commercetools payment to get a list of payment methods available to the shopper.

> - [Step 3](#step-3-add-components-to-your-payments-form): Set `getOriginKeysRequest` custom field to commercetools payment to get origin key and add the specific payment method Component to your checkout payments form.

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/04d6cc509cb9f0b7812cbc9b075fe621698644b0

- [Step 4](#step-4-make-a-payment): Submit a payment request with setting `makePaymentRequest` custom field with the payment data returned by the Component.
- [Step 5](#step-5-submit-additional-payment-details): Set `getPaymentMethodsRequest` custom field to commercetools payment to submit additional payment details. 
- [Step 6](#step-6-capture-payment-required-for-klarna): Add a `Charge` transaction to commercetools payment to manually capture payment.

> ## Before you begin

Author: It's mentioning that integration should be after deployment (with a running extension module) with Adyen keys etc. So it's more like referencing other documents and IMO should be here.

In this case, load order by ordered cart ID and show order confirmation page.
This might happen if the cart has been already ordered in a different tab 
or by asynchronous process like [commercetools-payment-to-order-processor job](https://github.com/commercetools/commercetools-payment-to-order-processor).

### Recalculate cart
[Execute cart recalculate](https://docs.commercetools.com/http-api-projects-carts#recalculate) to ensure:
 - Cart totals are always up-to-date 
 - Time-limited discounts are not removed/invalidated from the cart automatically. They are validated on recalculate and order creation only.

### Validate payment
There must be at least one commercetools payment object of type Adyen (`Payment.paymentMethodInfo.paymentInterface = ctp-adyen-integration`).

### Validate payment transaction
Cart's payment counts as successful if there is at least one payment object
with successful transaction state (`Payment.Transaction.state=Success`) 
and transactions type `Authorization` or `Charge`.

## Step 2: Get available payment methods

> When your shopper is ready to pay, get a list of the available payment methods based on their country, device, and the payment amount.

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/f05a6fe02c962ec3fd8fd0aa26f71cc5ec6c2a7a

> From the merchant server, [Create/Update commercetools payment](https://docs.commercetools.com/http-api-projects-payments#create-a-payment) with `getPaymentMethodsRequest` custom field.  

Author: Right, I removed the `From the merchant server` part, but it looks like I did not commit it separately: check this commit: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/ecb5ab87b229f2243044dd5993c52297ff5b6b44#diff-20d0a863a5962610863344eb769d92e3R95

> Set `getPaymentMethodsRequest` custom field for a shopper in the Germany, for a payment of `10 EUR`: 

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/ecb5ab87b229f2243044dd5993c52297ff5b6b44

    "postalCode": "12345",
    "street": "Stargatan"
  },
  "returnUrl": "https://www.your-company.com/...",
  "lineItems": [
    {
      "quantity": "1",
      "amountExcludingTax": "331",
      "taxPercentage": "2100",
      "description": "Shoes",
      "id": "Item #1",
      "taxAmount": "69",
      "amountIncludingTax": "400"
    },
    {
      "quantity": "2",
      "amountExcludingTax": "248",
      "taxPercentage": "2100",
      "description": "Socks",
      "id": "Item #2",
      "taxAmount": "52",
      "amountIncludingTax": "300"
    }
  ]
}
```
If `makePaymentRequest` has `lineItems` custom field, extension module will not overwrite those `lineItems`. This way you can also provide your own line item details if needed.

## Step 5: Submit additional payment details
If the shopper performed additional action (e.g. redirect) in the previous step,
you need to make `submitAdditionalPaymentDetailsRequest` to either complete the payment, or to check the payment result.

Collect information from the previous step and [update commercetools payment](https://docs.commercetools.com/http-api-projects-payments#update-payment) with `submitAdditionalPaymentDetailsRequest` custom field.

> The information is available either in `state.data.details` from the `onAdditionalDetails` event or, for redirects, the parameters you received when the shopper was redirected back to your website.

Author: Similar: https://github.com/commercetools/commercetools-adyen-integration/pull/329#discussion_r449214240

https://github.com/commercetools/commercetools-adyen- integration/commit/ecb5ab87b229f2243044dd5993c52297ff5b6b44

    },
    "fields": {
      "submitPaymentDetailsRequest": "{\"details\":{\"redirectResult\":\"Ab02b4c0!...\"}}",
      "submitAdditionalPaymentDetailsResponse": "{\"pspReference\":\"853592567856061C\",\"resultCode\":\"Authorised\",\"amount\":{\"currency\":\"EUR\",\"value\":1000},\"merchantReference\":\"YOUR_ORDER_REFERENCE\"}"
    }
  },
  "transactions": [
    {
      "id": "eab650fd-8616-471b-b884-eef641b4f169",
      "type": "Authorization",
      "amount": {
        "type": "centPrecision",
        "currencyCode": "EUR",
        "centAmount": 1000,
        "fractionDigits": 2
      },
      "interactionId": "853592567856061C",
      "state": "Success"
    }
  ]
}
```

## Step 6: Capture payment (required for Klarna)
All Klarna payments [have to be manually captured](https://docs.adyen.com/payment-methods/klarna/web-component#capture) within 28 days after authorisation, even if you have enabled automatic capture on your merchant account.
Refer to [Manual Capture](ManualCapture.md) guide to see how it can done.

## Error handling

In case you encounter errors in your integration, refer to the following:

### Extension module errors

> If you receive a `non-HTTP 200 response`, use the commercetools `interface interactions` to troubleshoot the request and errors.

Reviewer 2: ```suggestion
If you receive a `non-HTTP 200 response`, use the commercetools payment `interface interactions` to troubleshoot the response.
```

> Interface interactions can be requests sent to the Adyen, responses received from the Adyen or notifications received from the Adyen. 

Reviewer 2: ```suggestion
Interface interactions can represent a `request` sent to Adyen, a `response`, or a `notification` received from Adyen. 
```

> Some interactions may result in a transaction. If so, the interactionId in the Transaction should be set to match the `pspReference` of the Adyen for the interaction.

Reviewer 2: ```suggestion
Some interactions may result in a transaction. If so, the interactionId in the payment transaction will be set to match the `pspReference` of the Adyen API.
```

### Adyen payment refusals

If you receive an HTTP 200 response with an Error or Refused resultCode, check the refusal reason and, if possible, modify your request.

Check the following table to see the mapping of Adyen [result codes](https://docs.adyen.com/development-resources/response-handling#error-codes-types) to commercetools [transaction state](https://docs.commercetools.com/http-api-projects-payments#transactionstate)
|Adyen result code| The commercetools transaction state
| --- | --- |
| Authorised| Success|
| Refused| Failure|

> | Error| Failure|

Reviewer 2: In case of Error or Refused do we create a transaction at all? If not, it means that failure will not be set to the transaction?

Author: We are creating Authorization transaction with those states.

https://github.com/commercetools/commercetools-adyen-integration/blob/eb945e9e1fd373b2078e05cefe19573e7d0601c7/extension/src/paymentHandler/payment-utils.js#L96-L105

Author: https://github.com/commercetools/commercetools-adyen-integration/pull/329/commits/f95f269c2df4190109589c26117e8937312afa21

### Shopper successfully paid but `redirectUrl` was not reached
In some payment redirect cases, there might be a valid payment but no order as shopper did not reach the shop's `redirectUrl`.
For example, after successfully issued payment shopper loses internet connection or accidentally closes the tab.

> In this case [Notification module](../../notification) will receive later a notification with successful content, process it, and update the payment.

Reviewer 2: ```suggestion
In this case [Notification module](../../notification) will receive asynchronously a notification from Adyen with payment confirmation which will result in a transaction creation or transaction state change.
```

> Usage of scheduled [commercetools-payment-to-order-processor](https://github.com/commercetools/commercetools-payment-to-order-processor) job ensures that for every successful payment

Reviewer 2: ```suggestion
An optional usage of scheduled [commercetools-payment-to-order-processor](https://github.com/commercetools/commercetools-payment-to-order-processor) job ensures that for every successful payment
```

an order can still be asynchronously created.

### Shopper tries to pay a different amount than the actual order amount
For redirect payments payment amount is bound to `redirectUrl`.
After redirect and before the actual finalization of the payment at the provider's page, the shopper is still able to change the cart's amount within the second tab.
If shopper decides to change cart's amount within the second tab and finalize payment within the first tab, then according to payment amount validation an error

> will be shown and order creation must be declined.

Reviewer 2: ```suggestion
will be shown and order creation must be declined. In such a case, it might be reasonable to [cancel or refund](#cancel-or-refund) the invalid payment.
```

## Test and go live

> After doing the all steps above and testing the payment methods with your test accounts, then when you are ready to go live, you need to do the following steps described as [testing the integration](https://docs.adyen.com/checkout/components-web/#testing-your-integration).

Reviewer 2: ```suggestion
Before you go live please follow [steps](https://docs.adyen.com/checkout/components-web/#testing-your-integration) described by Adyen.
```

> Additionally, follow the official Adyen [integration checklist](https://docs.adyen.com/development-resources/integration-checklist) to ensure you have a complete implementation into Adyen.

Reviewer 2: ```suggestion
Additionally, follow the official Adyen [integration checklist](https://docs.adyen.com/development-resources/integration-checklist).
```

# Manual Capture

> By default, payments are captured immediately after authorisation. For payment methods that support separate authorisation and capture, you also have the option to capture the payment later, for example only after the goods have been shipped. This allows you to cancel the payment in case of any issues with the shipment. 

Reviewer 2: ```suggestion
By default, payments are captured immediately (or with [delay](https://docs.adyen.com/point-of-sale/capturing-payments/delayed-capture#set-up-delayed-capture)) after authorisation. For payment methods that support separate authorisation and capture, you also have the option to capture the payment later, for example only after the goods have been shipped. This also allows you to cancel the payment/authorisation.
```
